### PR TITLE
Jkb/49/50

### LINKED ISSUE #49: Improve the thumbnail algorithm

### LINKED ISSUE #50: Refactor EF configuration to make it cleaner

### PR DESCRIPTION
**Pull request by JamesBock**
closes #49 closes #50 
Notes: 
-Postgres prefers you use a package called NodaTime for any Date/Time operations beyond the most basic instances (which we might fall into). I kept getting an error when trying to add a migration when using it because we probably don't need the advanced support it gives for time zones at this point.
-Table and column names are all in snake_case to fit with PSQL conventions. Postgres "folds down" to lower case while EF Core uses the class name as is by default. Should prevent excessive qutation marking or queries
-Thumbnail algorithm is untested but theoretically crops the image to 125 x 100px image.

### CODE REVIEW

**Comment by JamesBock** [comment]
The Test confilct is from me commenting out a line because the tests were failing because the SmsController now takes a TenantContent in the constructor.

File: TenantFile/Api.UnitTests/SmsControllerTests.cs [hunk 1]
```diff
@@ -58,10 +58,10 @@ public async Task UploadsImageAndReturnsCorrectResponse()
                 .AddInMemoryCollection(myConfiguration)
                 .Build();
 
-            var controller = new SmsController(Mock.Of<ILogger<SmsController>>(), cloudStorageMock.Object, configuration)
-            {
-                ControllerContext = ctx
-            };
+            //var controller = new SmsController(Mock.Of<ILogger<SmsController>>(), cloudStorageMock.Object, configuration)
```
Comment (ryanrightmer) [inline TenantFile/Api.UnitTests/SmsControllerTests.cs:61]: I have a fix for this that I can merge to master.

File: TenantFile/Api/Controllers/SmsController.cs [hunk 2]
```diff
@@ -129,12 +129,20 @@ private async Task<IEnumerable<(string image, string thumbnail)>> SaveMedia(int
                 var httpClient = new HttpClient();
                 var response = await httpClient.GetAsync(mediaUrl);
                 var inputStream = await response.Content.ReadAsStreamAsync();
+                
                 using var image = SixLabors.ImageSharp.Image.Load(inputStream);
                 image.Mutate(x => x
-                     .Resize(0, 100));
+                     //.DetectEdges()
```
Comment (ryanrightmer) [inline TenantFile/Api/Controllers/SmsController.cs:135]: Unnecessary comment

File: TenantFile/Api/Controllers/SmsController.cs [hunk 3]
```diff
@@ -129,12 +129,20 @@ private async Task<IEnumerable<(string image, string thumbnail)>> SaveMedia(int
...
+                           Mode = ResizeMode.Crop,
+                           Size = new Size(125, 100)
+                           
+                        }));
+                
                 var outputStream = new MemoryStream();
-                image.Save(outputStream, new PngEncoder());
-
+                image.Save(outputStream, encoder: new PngEncoder() { CompressionLevel = PngCompressionLevel.BestSpeed });
+                image.Dispose();
```
Comment (ryanrightmer) [inline TenantFile/Api/Controllers/SmsController.cs:145]: Dispose isn't necessary, we're using a `using` statement on line 133 that will automatically clean up the resource.

File: TenantFile/Api/Models/Entities/ResidenceRecord.cs [hunk 4]
```diff
@@ -7,11 +7,11 @@ namespace TenantFile.Api.Models
 {
     public class ResidenceRecord
     {
-        [Key]
-        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+      
         public int Id { get; set; }
-        public DateTimeOffset MoveIn { get; set; }
-        public DateTimeOffset MoveOut { get; set; }
+        //NodaTime is the recommended way to interact with Postgres date/time types
```
Comment (ryanrightmer) [inline TenantFile/Api/Models/Entities/ResidenceRecord.cs:12]: I think we'll be fine for date time concerns as long as we follow the logic on this page: https://www.npgsql.org/doc/types/datetime.html

Let's go ahead and convert to DateTime, but not worry about nodatime for now.

File: TenantFile/Api/Models/TenantContext.cs [hunk 5]
```diff
@@ -1,55 +1,37 @@
...
+        public DbSet<Image> Images{ get; set; } = null!;
+        public DbSet<Address> Addresses { get; set; } = null!;
+        public DbSet<Residence> Residences{ get; set; } = null!;
+        public DbSet<ResidenceRecord> ResidenceRecords { get; set; } = null!;
 
-        protected override void OnConfiguring(DbContextOptionsBuilder options)
-        {
-            options.UseNpgsql(_connectionString);
-        }
+        //moved this to startup.cs but it could just as easily be here
```
Comment (ryanrightmer) [inline TenantFile/Api/Models/TenantContext.cs:24]: Yeah, I like what you've done better. Go ahead and remove this comment.

File: TenantFile/Api/TenantFile.Api.csproj [hunk 6]
```diff
@@ -8,6 +8,25 @@
     <UserSecretsId>2e183d23-5e31-4ab7-bd0d-d8ecef7a6760</UserSecretsId>
   </PropertyGroup>
   <ItemGroup>
+    <Compile Remove="Migrations\20200828160717_ToTableTenant.cs" />
```
Comment (ryanrightmer) [inline TenantFile/Api/TenantFile.Api.csproj:11]: I think your IDE probably added these. We want to keep the csproj clean, so let's remove them

File: TenantFile/Api/appsettings.json [hunk 7]
```diff
@@ -1,12 +1,18 @@
 {
+  "AllowedHosts": "*",
+  "CloudSQL": {
+    "ConnectionString": "Host=127.0.0.1;Uid=DATABASE_USER;Pwd=PASSWORD;Database=DATABASE_NAME"
+  },
+  "ConnectionStrings": {
+    "PostgresDBConection": "Host=localhost;Port=5432;Database=tenant_file;Username=postgres;Password=TenantFile"
```
Comment (ryanrightmer) [inline TenantFile/Api/appsettings.json:7]: This looks like a personal DB, so let's not check it in.

Comment (JamesBock) [reply TenantFile/Api/appsettings.json]: Removed local db

File: TenantFile/Api/TenantFile.Api.csproj (hunk 6 above)
Comment (JamesBock) [reply TenantFile/Api/TenantFile.Api.csproj]: Yeah this was something VS 2019 added. issue here https://github.com/dotnet/efcore/issues/18381 Should be safe to remove.

File: TenantFile/Api/Controllers/SmsController.cs (hunk 3 above)
Comment (JamesBock) [reply TenantFile/Api/Controllers/SmsController.cs]: Removed. I'm not used to not having the brackets with the cope of the using statement. =)

File: TenantFile/Api/Controllers/SmsController.cs (hunk 2 above)
Comment (JamesBock) [reply TenantFile/Api/Controllers/SmsController.cs]: gone

**Review by ryanrightmer — APPROVED** [review]
This is looking good to me

**Pull request merged**